File: sorting_algorithms/merge_sort/python/merge_sort.py
from cmath import inf
from math import floor

A = [5, 2, 4, 6, 1, 3] # unsorted array

def Merge(A, start, mid, end):
    n1 = mid - start + 1
    n2 = end - mid
    L = [] 
    R = [] 
    for i in range(1, n1 + 1):
        L.append(A[start + i - 1])
    for j in range(1, n2 + 1):
        R.append(A[mid + j])
    L.append(inf)
    R.append(inf)
    i = 0
    j = 0
    for k in range(start, end + 1):
        if L[i] <= R[j]:
            A[k] = L[i] # left wins
            i += 1
        else:
            A[k] = R[j] # right wins
            j += 1

def MergeSort(A, start, end):
    if start < end:
        mid = floor((start + end)/2)
        MergeSort(A, start, mid)
        MergeSort(A, mid + 1, end)
        Merge(A, start, mid, end)


MergeSort(A, 0, len(A) - 1)

print(list(A))

File: sorting_algorithms/merge_sort/TS/merge_sort.ts
function merge(A: Array<number>, start: number, mid: number, end: number, B: Array<number>) {
    let i = start
    let j = mid

    for (let k : number = start; k < end; k++) {
        if (i < mid && (j >= end || A[i] <= A[j])) {
            B[k] = A[i];
            i++;
        } else {
            B[k] = A[j];
            j++;
        }
    }
}

function mergeSort(B: Array<number>, start : number, end : number, A: Array<number>) {
    if (end - start <= 1) return;

    let mid = Math.floor((start + end) / 2);
    mergeSort(A, start, mid, B);
    mergeSort(A, mid, end, B);
    merge(B, start, mid, end, A);
}

// array of 5 ints 
let arr : Array<number> = [5, 4, 3, 2, 1];
let arr2 : Array<number> = arr.slice();

mergeSort(arr2, 0, arr.length, arr);
console.log(arr);